### PR TITLE
Issue #5412: vectorize diagnostic-only planner rollouts (cheap-lane worker)

### PR DESCRIPTION
**Pull request by ll7**
Refs #5412

## What / Why

This is a bounded successor slice for #5412 and does not duplicate the prior obstacle-feature vectorization work. It vectorizes three diagnostic-only (non-benchmark) planner adapters identified in #5412; benchmark-facing candidates remain separate numeric-parity-gated work.

### Changes

1. **CrowdNav_HEIGHT raycast** (`crowdnav_height.py`): replace the triple-nested Python loop (rays × segments × pedestrians) with NumPy broadcasting. Obstacle segment intersections use `(rays, segments)` parametric-line broadcast; pedestrian disc intersections use `(rays, pedestrians)` circle-line quadratic calculations. The vectorized path preserves the scalar helper's inclusive `t >= 0` boundary behavior and rejects non-finite intersections.

2. **SonicCrowdNav spatial edges** (`sonic_crowdnav.py`): replace the per-pedestrian prediction loop with `(pedestrians, timesteps, 2)` broadcasting and reshape/interleave.

3. **Learned risk surface** (`learned_risk_surface.py`): replace the per-pedestrian Gaussian bump loop with `(N_pedestrians, grid_h, grid_w)` broadcasting plus max reduction. The existing world-to-ego transform is applied in batch.

### Why this matters

These are high-impact Python loops in non-benchmark adapters. CrowdNav_HEIGHT's raycast avoids Python iterations over rays, obstacles, and pedestrians per planner step. The changes remain explicitly diagnostic-only and make no benchmark score, determinism, or numeric-parity claim.

### Validation

- `uv run --extra training pytest -q tests/planner/test_learned_risk_surface.py tests/planner/test_crowdnav_height.py tests/planner/test_sonic_crowdnav.py`: 29 passed, 6 skipped.
- The gate added regression coverage for multiple obstacle segments, zero-distance segment hits, and tangent pedestrian intersections.
- Ruff check and format check pass on all changed files.

### Issue #5412 remaining candidates

- DWA / RiskDWA velocity-lattice vectorization (benchmark-gated)
- MPPI / predictive-MPPI rollout vectorization (benchmark-gated)
- socnav PredictionPlannerAdapter rollout/broadcast (benchmark-gated)
- SocialForce vectorization (benchmark-gated)
- ORCA rvo2 persistent simulator (benchmark-gated)

No campaign, training, or benchmark result is claimed here.

## Summary by CodeRabbit

* **Performance**
  * Improved lidar sensing, risk-surface generation, and pedestrian prediction calculations for faster batch processing.
  * Preserved sensor range, visibility, sorting, and risk-surface behavior.

* **Bug Fixes**
  * Correctly handles multiple obstacle segments and zero-distance ray intersections, including boundary cases.

* **Tests**
  * Added coverage for broadcasted obstacle raycasting and zero-hit scenarios.

### CODE REVIEW

**Comment by ll7** [comment]
Gate phase 1 findings for PR #5497, reviewing the original implementation head `566ae1b54c6d2273ee69c2838a51b46733ff8fc4` and the current repaired branch:

Open review-thread enumeration (exhaustive):

1. Gemini `discussion_r3569998341` (`robot_sf/planner/crowdnav_height.py`): `u_num` used incompatible `(segments,)` and `(rays,)` shapes, and obstacle/pedestrian ray intersections excluded valid `t=0` boundary hits. Verdict: fixed in phase 2 with `(rays, segments)` broadcasting, finite-value guards, inclusive `t >= 0`, and regression coverage for multiple segments and zero-distance/tangent hits.
2. Gemini `discussion_r3569998369` (`robot_sf/planner/crowdnav_height.py`): pedestrian quadratic intersections used strict `t > 0` and changed the scalar range-sensor boundary semantics. Verdict: fixed in phase 2 with inclusive roots and a tangent-at-origin regression test.
3. Gemini `discussion_r3569998381` (`robot_sf/planner/sonic_crowdnav.py`): `np.stack` redundantly copied an already `(used, steps, 2)` array before reshape. Verdict: fixed in phase 2 with direct reshape.
4. Gate contract finding: the original body overclaimed identical numerical results for a diagnostic-only vectorization and lacked the required successor/non-duplication statement for issue #5412. Verdict: fixed in phase 2; the current body makes no benchmark/numeric-parity claim, states the bounded successor slice, and maps benchmark-facing candidates to follow-up work.
5. Gate staleness finding: the PR branch was based on `5cefcab23cf5e62e77c203f3d87be47fc3819c06`, behind current main after the batch merges. Verdict: fixed in phase 2 by merging current main `d87b77f622117177767535d54bd1a7d6d3c1c25a` before fresh CI.

CodeRabbit has 0 open inline review threads; its review was skipped. Gemini has 3 open threads (items 1–3). No finding is waived or spun off; all five findings are fixed in phase 2.

File: robot_sf/planner/crowdnav_height.py [hunk 1]
```diff
@@ -738,27 +736,61 @@ def _raycast_obstacles(
...
+                - directions[:, 1, None] * seg_vec[None, :, 0]
+            )
+            delta = seg_starts - origin  # (S, 2)
+            # t = (delta x seg) / denom, u = (delta x dir) / denom
+            t_num = delta[:, 0] * seg_vec[:, 1] - delta[:, 1] * seg_vec[:, 0]  # (S,)
+            u_num = delta[:, 0] * directions[:, 1] - delta[:, 1] * directions[:, 0]  # (R,)
+            with np.errstate(divide="ignore", invalid="ignore"):
+                t = t_num[None, :] / denom  # (R, S)
+                u = u_num[:, None] / denom  # (R, S)
+            valid = (t > 0) & (u >= 0) & (u <= 1) & (np.abs(denom) > 1e-9) & np.isfinite(t)
```
[the remark this answers is not in the document]
Comment (ll7) [reply robot_sf/planner/crowdnav_height.py]: Fixed in 5293ee5d0: `u_num` now broadcasts to `(rays, segments)`, and the scalar boundary semantics are preserved with inclusive `t >= 0` plus finite-value guards. The new regression test exercises multiple segments and a zero-distance hit.

File: robot_sf/planner/crowdnav_height.py [hunk 2]
```diff
@@ -738,27 +736,61 @@ def _raycast_obstacles(
...
+                directions[:, 0, None] * p1x[None, :] + directions[:, 1, None] * p1y[None, :]
+            )  # (R, P)
+            c = norm_p1[None, :] - r_sq  # (R, P)
+            disc = b**2 - 4.0 * a[:, None] * c  # (R, P)
+            with np.errstate(divide="ignore", invalid="ignore"):
+                sqrt_disc = np.sqrt(np.maximum(disc, 0.0))
+                t1 = (-b - sqrt_disc) / (2.0 * a[:, None])
+                t2 = (-b + sqrt_disc) / (2.0 * a[:, None])
+            valid_t = (disc >= 0) & ((t1 > 0) | (t2 > 0)) & np.isfinite(t1) & np.isfinite(t2)
+            pos_t = np.where(t1 > 0, t1, t2)
```
[the remark this answers is not in the document]
Comment (ll7) [reply robot_sf/planner/crowdnav_height.py]: Fixed in 5293ee5d0: pedestrian quadratic roots now use inclusive `t >= 0`, matching the shared scalar range-sensor helper. The tangent-at-origin regression test covers the boundary.

File: robot_sf/planner/sonic_crowdnav.py [hunk 3]
```diff
@@ -462,15 +462,21 @@ def _build_model_inputs(
...
+            rel = np.asarray(ped_positions[:used], dtype=float) - np.asarray(
+                robot_pos, dtype=float
+            )  # (used, 2)
+            vel = np.asarray(ped_velocities[:used], dtype=float)  # (used, 2)
+            steps = np.arange(predict_steps + 1, dtype=float).reshape(1, -1, 1) * float(
+                time_step
+            )  # (1, predict_steps+1, 1)
+            future = rel[:, None, :] + vel[:, None, :] * steps
+            # Interleave (x,y) pairs into flat sequences: (used, 2*(predict_steps+1))
+            seq = np.stack([future[:, :, 0], future[:, :, 1]], axis=-1).reshape(used, -1)
```
[the remark this answers is not in the document]
Comment (ll7) [reply robot_sf/planner/sonic_crowdnav.py]: Fixed in 5293ee5d0: `future` already has the required `(used, predict_steps + 1, 2)` layout, so the redundant `np.stack` was removed in favor of a direct reshape.

**Comment by ll7** [comment]
Gate phase 2 repair verification for PR #5497 at head `c73d055239a1896f3a80e293293991fbb9ab3f6`:

Open review-thread enumeration (exhaustive):

1. Gemini `discussion_r3569998341`: ray/segment `u_num` shape mismatch and non-inclusive ray boundary. Verdict: fixed in `5293ee5d0`; broadcast shape, finite guards, inclusive `t >= 0`, and multi-segment/zero-hit tests are present.
2. Gemini `discussion_r3569998369`: pedestrian quadratic strict-positive boundary. Verdict: fixed in `5293ee5d0`; inclusive roots and tangent-at-origin coverage are present.
3. Gemini `discussion_r3569998381`: redundant Sonic `np.stack`. Verdict: fixed in `5293ee5d0` with direct reshape.
4. Gate body/claim finding: overclaim and missing successor statement. Verdict: fixed in the current body; it explicitly scopes this as a diagnostic successor slice, makes no benchmark/numeric-parity claim, and maps benchmark-facing candidates to follow-up work.
5. Gate staleness finding: branch was behind current main. Verdict: fixed by merge commit `c73d0552`, based on current main `d87b77f6`.

There are 0 open review threads. Gemini: 0; CodeRabbit: 0 open inline threads (review skipped). No finding is waived or spun off.

Local proof:

- `uv run --extra training pytest -q tests/planner/test_learned_risk_surface.py tests/planner/test_crowdnav_height.py tests/planner/test_sonic_crowdnav.py`: 29 passed, 6 skipped.
- Ruff check and format check on all changed files: clean.
- The new regression coverage exercises multiple segments, zero-distance segment hits, and tangent pedestrian intersections.

Fresh CI is pending on this current head; merge remains gated on all configured checks and a CLEAN, up-to-date branch.

**Comment by ll7** [comment]
Gate phase 3 checklist for PR #5497 at head `c73d055239a1896f3a80e293293991fbb9ab3f6`:

- Findings enumeration is recorded in phase 1/2 comments: 0 open review threads. Gemini: 0; CodeRabbit: 0 open inline threads (review skipped). All three bot findings plus the two gate findings are fixed; none is waived or spun off.
- The current body is authoritative: bounded successor slice for #5412, no duplicate claim, diagnostic-only scope, no benchmark/numeric-parity claim, and follow-up candidates listed.
- The branch includes current main `d87b77f622117177767535d54bd1a7d6d3c1c25a`; direct PR state is non-draft, `mergeable_state=clean`, `mergeable=true`.
- All configured checks pass: Python analysis, CodeQL, compatibility matrix, fast feedback, examples smoke, wheel smoke, promoted-planner smoke, artifact/xdist smoke, Ruff/security, type advisory, CodeRabbit routing, contract checks, and aggregate `ci`.
- Local focused proof: 29 passed, 6 skipped with `uv run --extra training`; Ruff check/format check clean.
- Main CI is separately red at run `29242185072` on `9dbcbdd6a`; per the 2026-07-13 policy this is advisory only. The suspected pinned-hash breakage was repaired by #5494, and #5497 does not touch those files.

Verdict: merge-ready.
